File: src-node/services/scoring.js
function numericEmployeeCount(value) {
  const parsed = Number.parseInt(value, 10);
  return Number.isNaN(parsed) ? 0 : parsed;
}

function scoreBudget(lead) {
  let score = 0;
  const country = (lead.country || "").toLowerCase();
  const employees = numericEmployeeCount(lead.employee_count);
  const fundingStage = (lead.funding_stage || "").toLowerCase();

  if (["united states", "usa", "us", "united kingdom", "uk"].includes(country)) {
    score += 15;
  }
  if (employees >= 15 && employees <= 60) {
    score += 20;
  } else if ((employees >= 8 && employees < 15) || (employees >= 61 && employees <= 100)) {
    score += 10;
  }
  if (["seed", "series a", "series b", "private"].includes(fundingStage)) {
    score += 15;
  }
  if (!country && /\.[a-z]{2,}$/.test((lead.email || "").toLowerCase())) {
    score += 5;
  }
  if (!employees && (lead.company_name || "").trim()) {
    score += 5;
  }
  return score;
}

function scoreUrgency(lead) {
  const text = `${lead.recent_signal || ""} ${lead.hiring_signal || ""}`.toLowerCase();
  let score = 0;

  if (["funding", "raised", "growth", "expanding"].some((token) => text.includes(token))) {
    score += 15;
  }
  if (
    ["sales ops", "revops", "operations", "customer success", "marketing ops"].some((token) =>
      text.includes(token)
    )
  ) {
    score += 15;
  }
  if (["hiring", "job post", "opening"].some((token) => text.includes(token))) {
    score += 10;
  }
  return score;
}

function scoreFit(lead) {
  const text = `${lead.industry || ""} ${lead.tech_stack || ""} ${lead.pain_notes || ""}`.toLowerCase();
  let score = 0;

  if (["saas", "law", "legal", "real estate"].some((token) => text.includes(token))) {
    score += 10;
  }
  if (["hubspot", "salesforce", "slack", "whatsapp", "crm", "zapier"].some((token) => text.includes(token))) {
    score += 15;
  }
  if (
    ["manual", "slow", "spreadsheet", "handoff", "inbound", "intake", "follow-up"].some((token) =>
      text.includes(token)
    )
  ) {
    score += 20;
  }
  if (
    ["automation", "implement", "integration", "migration", "operations", "workflow", "redlining"].some((token) =>
      text.includes(token)
    )
  ) {
    score += 15;
  }
  return score;
}

function scoreBuyer(lead) {
  const title = (lead.buyer_title || "").toLowerCase();
  if (["job application", "jobs", "applicant"].some((token) => title.includes(token))) {
    return -8;
  }
  if (["ceo", "founder", "coo", "cro", "cto", "head"].some((token) => title.includes(token))) {
    return 15;
  }
  if (["director", "manager", "lead"].some((token) => title.includes(token))) {
    return 8;
  }
  if (["consultant", "engineer", "specialist", "operations", "success"].some((token) => title.includes(token))) {
    return 5;
  }
  return 0;
}

export function scoreLead(lead) {
  const budget_score = scoreBudget(lead);
  const urgency_score = scoreUrgency(lead);
  const fit_score = scoreFit(lead);
  const buyer_score = scoreBuyer(lead);
  const title = (lead.buyer_title || "").toLowerCase();
  const qualificationNotes = [];

  if (["job application", "jobs", "applicant"].some((token) => title.includes(token))) {
    qualificationNotes.push("Contact looks job-post or applicant oriented, not a clean buyer profile.");
  }
  if (["ceo", "founder", "coo", "cro", "cto", "head", "director", "manager", "lead"].some((token) => title.includes(token))) {
    qualificationNotes.push("Role shows some seniority signal.");
  }
  if ((lead.industry || "").toLowerCase().includes("enterprise")) {
    qualificationNotes.push("Enterprise or operations-heavy context may fit process automation.");
  }

  return {
    budget_score,
    urgency_score,
    fit_score,
    buyer_score,
    lead_score: budget_score + urgency_score + fit_score + buyer_score,
    qualification_notes: qualificationNotes.join(" ")
  };
}

File: src/scoring.py
from __future__ import annotations

from dataclasses import dataclass

from lead_model import Lead


@dataclass
class LeadScore:
    total: int
    budget_score: int
    urgency_score: int
    fit_score: int
    buyer_score: int


def score_lead(lead: Lead) -> LeadScore:
    budget_score = score_budget(lead)
    urgency_score = score_urgency(lead)
    fit_score = score_fit(lead)
    buyer_score = score_buyer(lead)
    total = budget_score + urgency_score + fit_score + buyer_score
    return LeadScore(
        total=total,
        budget_score=budget_score,
        urgency_score=urgency_score,
        fit_score=fit_score,
        buyer_score=buyer_score,
    )


def score_budget(lead: Lead) -> int:
    score = 0

    if lead.country.lower() in {"united states", "usa", "us", "united kingdom", "uk"}:
        score += 15
    if 15 <= lead.employee_count <= 60:
        score += 20
    elif 8 <= lead.employee_count < 15 or 61 <= lead.employee_count <= 100:
        score += 10
    if lead.funding_stage.lower() in {"seed", "series a", "series b", "private"}:
        score += 15

    return score


def score_urgency(lead: Lead) -> int:
    text = f"{lead.recent_signal} {lead.hiring_signal}".lower()
    score = 0

    if any(keyword in text for keyword in ("funding", "raised", "growth", "expanding")):
        score += 15
    if any(keyword in text for keyword in ("sales ops", "revops", "operations", "customer success", "marketing ops")):
        score += 15
    if any(keyword in text for keyword in ("hiring", "job post", "opening")):
        score += 10

    return score


def score_fit(lead: Lead) -> int:
    text = f"{lead.industry} {lead.tech_stack} {lead.pain_notes}".lower()
    score = 0

    if any(keyword in text for keyword in ("saas", "law", "legal", "real estate")):
        score += 10
    if any(keyword in text for keyword in ("hubspot", "salesforce", "slack", "whatsapp", "crm", "zapier")):
        score += 15
    if any(keyword in text for keyword in ("manual", "slow", "spreadsheet", "handoff", "inbound", "intake", "follow-up")):
        score += 20

    return score


def score_buyer(lead: Lead) -> int:
    title = lead.buyer_title.lower()
    if any(keyword in title for keyword in ("ceo", "founder", "coo", "cro", "cto", "head")):
        return 15
    if any(keyword in title for keyword in ("director", "manager", "lead")):
        return 8
    return 0
